src/TypeShape/StagingExtensions.fs

```diff
@@ -135,14 +128,13 @@ module Expr =
             aux <@ invalidOp "invalid tag" @> (cases.Length - 1))
 
     /// traverses an expression tree applying the transformation
-    /// `(fun x y z .. -> M[x,y,z,..]) a b c => M[a,b,c,..]`,
-    /// where a,b,c are variables or constants
+    /// `(fun x y z .. -> M[x,y,z,..]) a b c` => `M[a,b,c,..]`
     let unlambda (expr : Expr<'T>) =
         let (|AppLambdas|_|) (e : Expr) =
             // traverses the "App(App(App ... " part of the expression
             let rec gatherApps args e =
                 match e with
-                | Application(lhs, ((Var _ | Value _) as rhs)) -> gatherApps (rhs :: args) lhs
+                | Application(lhs, rhs) -> gatherApps (rhs :: args) lhs
```

Comment:
Why the change? If the right-hand-side expression is impure then this is not a valid reduction.

Reply:
Yes, agreed. But this reduction is useful in some scenario.
For example:
```fsharp
type Foo = { Value : int; Name: string }
let x  = Var("x"; typeof<Foo>) |> Expr.Var |> Expr.Cast<Foo>

let eq = <@ (=) @>
let const = <@ 1 @>
let expr1 = <@ (%eq) (%x).Value %c  @>

let startsWith = <@ fun (x: string) (y: string) -> x.StartsWith y @>
let substr  = <@ "abc" @>
let expr2 = <@ (%startsWith) (%x).Name %substr @> 
```
Or in my code: [declaration](https://github.com/habib-sadullaev/Equal/blob/81a991961115bfe3caf7bfa4614458a50251845c/src/Equal/Operators.fs#L16) and [usage](https://github.com/habib-sadullaev/Equal/blob/master/src/Equal/Expression.fs#L96)

Reply:
It violates the evaluation semantics of F# and would be a breaking change. I therefore would not be able to accept this even if it could be meaningful in certain scenaria.

Reply:
how about adding a new function, say ``unlambdaForced`` instead of changing ``unlambda``?

Reply:
It seems like a fairly straightforward variation to include as a helper function in your code.

Reply:
yes, I already did this). what about my other changes, if I revert this one?

Reply:
I think replacing the `Array.fold[Back]` calls with `Array.reduce[Back]` are good improvements.